File: modules/vescape-core/ios/geo/GeoMath.swift
import Foundation

/// Great-circle helpers shared by everything native that reasons about two positions.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/geo/GeoMath.kt
internal enum GeoMath {
  private static let earthRadiusMeters = 6_371_000.0

  static func distanceMeters(
    _ fromLatitude: Double,
    _ fromLongitude: Double,
    _ toLatitude: Double,
    _ toLongitude: Double
  ) -> Double {
    let deltaLatitude = (toLatitude - fromLatitude) * .pi / 180
    let deltaLongitude = (toLongitude - fromLongitude) * .pi / 180
    let a = sin(deltaLatitude / 2) * sin(deltaLatitude / 2)
      + cos(fromLatitude * .pi / 180) * cos(toLatitude * .pi / 180)
      * sin(deltaLongitude / 2) * sin(deltaLongitude / 2)
    return 2 * earthRadiusMeters * asin(min(1.0, sqrt(a)))
  }

  /// Initial great-circle bearing from one position to another, normalized to `[0, 360)`.
  static func travelBearingDeg(
    _ fromLatitude: Double,
    _ fromLongitude: Double,
    _ toLatitude: Double,
    _ toLongitude: Double
  ) -> Double {
    let fromLatitudeRad = fromLatitude * .pi / 180
    let toLatitudeRad = toLatitude * .pi / 180
    let deltaLongitude = (toLongitude - fromLongitude) * .pi / 180
    let y = sin(deltaLongitude) * cos(toLatitudeRad)
    let x = cos(fromLatitudeRad) * sin(toLatitudeRad)
      - sin(fromLatitudeRad) * cos(toLatitudeRad) * cos(deltaLongitude)
    return normalizeBearingDeg(atan2(y, x) * 180 / .pi) ?? 0
  }

  /// `[0, 360)`, or nil when the value is absent or not finite.
  static func normalizeBearingDeg(_ value: Double?) -> Double? {
    guard let value, value.isFinite else { return nil }
    return ((value.truncatingRemainder(dividingBy: 360)) + 360)
      .truncatingRemainder(dividingBy: 360)
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/geo/GeoMath.kt
package expo.modules.vescapecore.geo

import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Great-circle helpers shared by everything native that reasons about two positions.
 *
 * @parity /modules/vescape-core/ios/geo/GeoMath.swift
 */
object GeoMath {
  private const val EARTH_RADIUS_METERS = 6_371_000.0

  fun distanceMeters(
    fromLatitude: Double,
    fromLongitude: Double,
    toLatitude: Double,
    toLongitude: Double,
  ): Double {
    val deltaLatitude = Math.toRadians(toLatitude - fromLatitude)
    val deltaLongitude = Math.toRadians(toLongitude - fromLongitude)
    val a = sin(deltaLatitude / 2) * sin(deltaLatitude / 2) +
      cos(Math.toRadians(fromLatitude)) * cos(Math.toRadians(toLatitude)) *
      sin(deltaLongitude / 2) * sin(deltaLongitude / 2)
    return 2 * EARTH_RADIUS_METERS * asin(min(1.0, sqrt(a)))
  }

  /** Initial great-circle bearing from one position to another, normalized to `[0, 360)`. */
  fun travelBearingDeg(
    fromLatitude: Double,
    fromLongitude: Double,
    toLatitude: Double,
    toLongitude: Double,
  ): Double {
    val fromLatitudeRad = Math.toRadians(fromLatitude)
    val toLatitudeRad = Math.toRadians(toLatitude)
    val deltaLongitude = Math.toRadians(toLongitude - fromLongitude)
    val y = sin(deltaLongitude) * cos(toLatitudeRad)
    val x = cos(fromLatitudeRad) * sin(toLatitudeRad) -
      sin(fromLatitudeRad) * cos(toLatitudeRad) * cos(deltaLongitude)
    return normalizeBearingDeg(Math.toDegrees(atan2(y, x)))!!
  }

  /** `[0, 360)`, or null when the value is absent or not finite. */
  fun normalizeBearingDeg(value: Double?): Double? {
    if (value == null || !value.isFinite()) return null
    return ((value % 360) + 360) % 360
  }
}
